cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/EntityOperatorReconciler.java

```diff
@@ -117,13 +122,38 @@ public Future<Void> reconcile(boolean isOpenShift, ImagePullPolicy imagePullPoli
                 .compose(i -> userOperatorRoleBindings())
                 .compose(i -> topicOperatorConfigMap())
                 .compose(i -> userOperatorConfigMap())
+                .compose(i -> topicOperatorCruiseControlApiSecret())
                 .compose(i -> deleteOldEntityOperatorSecret())
                 .compose(i -> topicOperatorSecret(clock))
                 .compose(i -> userOperatorSecret(clock))
                 .compose(i -> deployment(isOpenShift, imagePullPolicy, imagePullSecrets))
                 .compose(i -> waitForDeploymentReadiness());
     }
 
+    /**
+     * Manages the Cruise Control API keys secret used by the Topic Operator.
+     *
+     * @return Future which completes when the reconciliation is done.
+     */
+    protected Future<Void> topicOperatorCruiseControlApiSecret() {
+        if (unidirectionalTopicOperator && isCruiseControlEnabled) {
+            String ccApiSecretName = KafkaResources.entityTopicOperatorCcApiSecretName(reconciliation.name());
+            if (entityOperator != null && entityOperator.topicOperator() != null) {
+                return secretOperator.getAsync(reconciliation.namespace(), ccApiSecretName)
+                    .compose(oldSecret -> {
+                        Secret newSecret = entityOperator.topicOperator().generateCruiseControlApiSecret(oldSecret);
+                        this.ccApiSecretHash = ReconcilerUtils.hashSecretContent(newSecret);
+                        return secretOperator.reconcile(reconciliation, reconciliation.namespace(), ccApiSecretName, newSecret)
+                            .map((Void) null);
+                    });
+            } else {
+                return secretOperator.reconcile(reconciliation, reconciliation.namespace(), ccApiSecretName, null)
+                    .map((Void) null);
+            }
+        }
+        return Future.succeededFuture();
+    }
+
```

Comment:
Wouldn't it be easier if we let the CruiseControlReconciler to handle the API credentials for the Topic Operator in a single API credentials secret? Is there any specific reason to have a separate secret for the CC API credentials for the Topic Operator?

Reply:
This was discussed before. We want to keep additional credentials isolated and close to the component using them, which adds some complexity, but it's definitely better in terms of user experience.

Reply:
They should have each its own secrets which they use even if the data have to be eventually shared in order to have the user configured in CC.

Reply:
> This was discussed before. We want to keep additional credentials isolated and close to the component using them, which adds some complexity, but it's definitely better in terms of user experience.

> They should have each its own secrets which they use even if the data have to be eventually shared in order to have the user configured in CC.

Based on the reasoning on this thread I do not agree and don't understand why we are doing this. That being said, it sounds like there was a previous thread (which has either been wiped or I just cannot find) that found this approach as the best way forward. Given I am late to the conversation and trust you both, I will defer to your judgement.

Reply:
@kyguy my initial approach to manage the TO credentials was like you say, which is also much easier to implement. Wit that, part of the TO configuration (API credentials) would be managed by the CC component, and TO should watch CC API secret and restart if its own credentials changed. Instead, with the current approach, every component exclusively manages its own API credentials, and CC has the extra responsibility of checking all API secrets (just TO at the moment) and integrate changes in its own API secret, which contains the authentication file used at runtime. Hope this clarifies a bit.


The real problem is that you can't detect API credentials change (e.g. secret recreation) from TO reconciler without reacting to CC API secret, which is a resource managed by another component.